File: code/testLineSegment.cpp
#include<iostream>
#include "LineSegment.hpp"
#include "unitTest.hpp"


void test_equals(){
    std::cout << "test_equals" << std::endl;
    bool result = true;

    Vector3D a(1.,0.,0.);
    Vector3D b(1.,0.,0.);
    
    result = (result && unitTest::assertTrue(a==b,"should be equal."));

    Vector3D c(0.,0.,0.);
    Vector3D d(0.,0.,0.);
    
    result = (result && unitTest::assertTrue(c==d,"should be equal."));

    
    if (!result) {
        std::cerr << "test_CrossProduct FAILED" << std::endl;
    }

    std::cout << std::endl;
}

void test_CrossProduct(){
    std::cout << "test_CrossProduct" << std::endl;
    bool result = true;

    Vector3D a(1.,0.,0.);
    Vector3D b(0.,1.,0.);
    
    result = (result && unitTest::assertTrue(a.cross(b)==Vector3D(0.,0.,1.0),"should be equal."));

    Vector3D c(1.,0.,0.);
    Vector3D d(1.,0.,0.);
    
    result = (result && unitTest::assertTrue(c.cross(d)==Vector3D(0.,0.,0.0),"should be equal."));

    
    if (!result) {
        std::cerr << "test_CrossProduct FAILED" << std::endl;
    }

    std::cout << std::endl;
}


void test_LineSegmentpositionDistance(){
    std::cout << "test_LineSegmentpositionDistance" << std::endl;
    bool result = true;

    Vector3D R1(0.,0.,0.);
    Vector3D R2(1.,0.,0.);

    LineSegment LineSegment1(R1,0.0,10.);
    LineSegment LineSegment2(R2,0.0,10.);


    result = (result && unitTest::assertMaxDiffence(LineSegment1.positionDistance(LineSegment2),1.0,1E-12,"positionDistance should be 1."));
    
    LineSegment2.change_position({1.,1.,0.});
    result = result && unitTest::assertMaxDiffence(LineSegment1.positionDistance(LineSegment2),sqrt(2.),1E-12,"positionDistance should be sqrt(2).");

    LineSegment2.change_position({1.,1.,1.});
    result = result && unitTest::assertMaxDiffence(LineSegment1.positionDistance(LineSegment2),sqrt(3.),1E-12,"positionDistance should be sqrt(3).");

    LineSegment2.change_position({0.,0.,0.});
    result = result && unitTest::assertMaxDiffence(LineSegment1.positionDistance(LineSegment2),0.,1E-12,"positionDistance should be 0.");

    LineSegment2.change_position({-1.,0.,0.});
    result = result && unitTest::assertMaxDiffence(LineSegment1.positionDistance(LineSegment2),1.,1E-12,"positionDistance should be 1.");

    LineSegment2.change_position({-10.,0.,0.});
    result = result && unitTest::assertMaxDiffence(LineSegment1.positionDistance(LineSegment2),10.,1E-12,"positionDistance should be 10.");

    LineSegment2.change_position({-INFINITY,0.,0.});
    result = result && unitTest::assertEqual(LineSegment1.positionDistance(LineSegment2),INFINITY,"positionDistance should be INFINITY.");

    if (!result) {
        std::cerr << "test_LineSegmentpositionDistance FAILED" << std::endl;
    }

    std::cout << std::endl;
}


void test_LineSegmentOverlap(){
    std::cout << "test_LineSegmentOverlap" << std::endl;
    bool result = true;

    LineSegment LineSegment1(Vector3D(0.,0.,0.),0.0,10.);
    LineSegment LineSegment2(Vector3D(1.,0.,0.),0.0,10.);

    result = (result && unitTest::assertFalse(LineSegment1.overlap(LineSegment2),"parallel shoult not overlap."));
    result = (result && unitTest::assertFalse(LineSegment2.overlap(LineSegment2),"parallel shoult not overlap."));

    LineSegment LineSegment3(Vector3D(0.,0.,0.),0.0,10.);
    LineSegment LineSegment4(Vector3D(0.,0.,0.),0.5*M_PI,10.);

    result = (result && unitTest::assertTrue(LineSegment3.overlap(LineSegment4),"cross shoult overlap."));
    result = (result && unitTest::assertTrue(LineSegment4.overlap(LineSegment3),"cross shoult overlap."));

    LineSegment LineSegment5(Vector3D(0.,0.,0.),0.0,10.);
    LineSegment LineSegment6(Vector3D(1000.,0.,0.),0.5*M_PI,10.);

    result = (result && unitTest::assertFalse(LineSegment5.overlap(LineSegment6),"cross far apart shoult not overlap."));
    result = (result && unitTest::assertFalse(LineSegment6.overlap(LineSegment5),"cross far apart shoult not overlap."));

    LineSegment LineSegment7(Vector3D(0.,0.,0.),0.0,100.);
    LineSegment LineSegment8(Vector3D(1.,0.,0.),0.6*M_PI,100.);

    result = (result && unitTest::assertTrue(LineSegment7.overlap(LineSegment8),"slanted cross shoult overlap."));
    result = (result && unitTest::assertTrue(LineSegment8.overlap(LineSegment7),"slanted cross shoult overlap."));

    LineSegment LineSegment9(Vector3D(0.,0.,0.),0.0,10000.);
    LineSegment LineSegment10(Vector3D(0.,1.,0.),0.1*M_PI,1000.);

    result = (result && unitTest::assertTrue(LineSegment9.overlap(LineSegment10),"almost parallel shoult overlap."));
    result = (result && unitTest::assertTrue(LineSegment10.overlap(LineSegment9),"almost parallel shoult overlap."));

    
    if (!result) {
        std::cerr << "test_LineSegmentOverlap FAILED" << std::endl;
    }

    std::cout << std::endl;
}





int main() {
    std::cout << std::endl;
    std::cout << "//-----------------------------------" << std::endl;
    std::cout << "STARTING SPHEROCYLINDER UNIT TESTS MAIN" << std::endl;
    std::cout << "//-----------------------------------" << std::endl;
    std::cout << std::endl;

    test_equals();
    test_CrossProduct();
    test_LineSegmentpositionDistance();
    test_LineSegmentOverlap();


    return 0;


}

File: code/python/testLineSegment.py
import unittest
import math
from LineSegment import Vector3D
from LineSegment import LineSegment

class TestVector3D(unittest.TestCase):

    def test_equals(self):
        a = Vector3D(1., 0., 0.)
        b = Vector3D(1., 0., 0.)
        self.assertEqual(a, b)

        c = Vector3D(0., 0., 0.)
        d = Vector3D(0., 0., 0.)
        self.assertEqual(c, d)

    def test_cross_product(self):
        a = Vector3D(1., 0., 0.)
        b = Vector3D(0., 1., 0.)
        self.assertEqual(a.cross(b), Vector3D(0., 0., 1.))

        c = Vector3D(1., 0., 0.)
        d = Vector3D(1., 0., 0.)
        self.assertEqual(c.cross(d), Vector3D(0., 0., 0.))

class TestLineSegment(unittest.TestCase):

    def test_position_distance(self):
        R1 = Vector3D(0., 0., 0.)
        R2 = Vector3D(1., 0., 0.)
        LineSegment1 = LineSegment(R1, 0.0, 10.)
        LineSegment2 = LineSegment(R2, 0.0, 10.)

        self.assertAlmostEqual(LineSegment1.position_distance(LineSegment2), 1.0, places=12)

        LineSegment2.change_position(Vector3D(1., 1., 0.))
        self.assertAlmostEqual(LineSegment1.position_distance(LineSegment2), math.sqrt(2), places=12)

        LineSegment2.change_position(Vector3D(1., 1., 1.))
        self.assertAlmostEqual(LineSegment1.position_distance(LineSegment2), math.sqrt(3), places=12)

        LineSegment2.change_position(Vector3D(0., 0., 0.))
        self.assertAlmostEqual(LineSegment1.position_distance(LineSegment2), 0., places=12)

        LineSegment2.change_position(Vector3D(-1., 0., 0.))
        self.assertAlmostEqual(LineSegment1.position_distance(LineSegment2), 1., places=12)

        LineSegment2.change_position(Vector3D(-10., 0., 0.))
        self.assertAlmostEqual(LineSegment1.position_distance(LineSegment2), 10., places=12)


    def test_line_segment_overlap(self):
        # Test case 1: Parallel segments should not intersect
        LineSegment1 = LineSegment(Vector3D(0., 0., 0.), 0.0, 10.)
        LineSegment2 = LineSegment(Vector3D(1., 0., 0.), 0.0, 10.)
        self.assertFalse(LineSegment1.intersect(LineSegment2))
        self.assertFalse(LineSegment2.intersect(LineSegment1))

        # Test case 2: Identical segments should not intersect
        self.assertFalse(LineSegment2.intersect(LineSegment2))

        # Test case 3: Perpendicular segments should intersect
        LineSegment3 = LineSegment(Vector3D(0., 0., 0.), 0.0, 10.)
        LineSegment4 = LineSegment(Vector3D(0., 0., 0.), 0.5 * math.pi, 10.)
        self.assertTrue(LineSegment3.intersect(LineSegment4))
        self.assertTrue(LineSegment4.intersect(LineSegment3))

        # Test case 4: Segments far apart should not intersect
        LineSegment5 = LineSegment(Vector3D(0., 0., 0.), 0.0, 10.)
        LineSegment6 = LineSegment(Vector3D(1000., 0., 0.), 0.5 * math.pi, 10.)
        self.assertFalse(LineSegment5.intersect(LineSegment6))
        self.assertFalse(LineSegment6.intersect(LineSegment5))

        # Test case 5: Slanted cross should intersect
        LineSegment7 = LineSegment(Vector3D(0., 0., 0.), 0.0, 100.)
        LineSegment8 = LineSegment(Vector3D(1., 0., 0.), 0.6 * math.pi, 100.)
        self.assertTrue(LineSegment7.intersect(LineSegment8))
        self.assertTrue(LineSegment8.intersect(LineSegment7))

        # Test case 6: Almost parallel, large segments should intersect
        LineSegment9 = LineSegment(Vector3D(0., 0., 0.), 0.0, 10000.)
        LineSegment10 = LineSegment(Vector3D(0., 1., 0.), 0.1 * math.pi, 1000.)
        self.assertTrue(LineSegment9.intersect(LineSegment10))
        self.assertTrue(LineSegment10.intersect(LineSegment9))

if __name__ == '__main__':
    unittest.main()
